services/packages/arch/verificator.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package arch
+
+import (
+	"errors"
+	"fmt"
+
+	"code.gitea.io/gitea/models/asymkey"
+	"code.gitea.io/gitea/models/db"
+	org "code.gitea.io/gitea/models/organization"
+	"code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/context"
+	"github.com/ProtonMail/gopenpgp/v2/crypto"
+)
+
+type IdentidyOwnerParameters struct {
+	*context.Context
+	Owner string
+	Email string
+}
+
+// This function will find user related to provided email adress and check if
+// he is able to push packages to provided namespace (user/organization/or
+// empty namespace allowed for admin users). Function will return user making
+// operation, organization or user owning the package.
+func IdentifyOwner(ctx *context.Context, owner, email string) (*user.User, *org.Organization, error) {
+	u, err := user.GetUserByEmail(ctx, email)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to find user with email %s, %v", email, err)
+	}
+
+	if owner == "" && u.IsAdmin {
+		return u, (*org.Organization)(u), nil
+	}
+
+	if owner == u.Name {
+		return u, (*org.Organization)(u), nil
+	}
+
+	if u.Name != owner {
+		org, err := org.GetOrgByName(ctx, owner)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to get organization: %s, %v", owner, err)
+		}
+		ismember, err := org.IsOrgMember(u.ID)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to check if user %s belongs to organization %s: %v", u.Name, org.Name, err)
+		}
+		if !ismember {
+			return nil, nil, fmt.Errorf("user %s is not member of organization %s", u.Name, org.Name)
+		}
+		return u, org, nil
+	}
+	return nil, nil, fmt.Errorf("unknown package owner")
+}
+
+// Validate package signature with owner's GnuPG keys stored in gitea's database.
+func ValidatePackageSignature(ctx *context.Context, pkg, sign []byte, u *user.User) error {
+	keys, err := asymkey.ListGPGKeys(ctx, u.ID, db.ListOptions{
+		ListAll: true,
+	})
+	if err != nil {
+		return errors.New("unable to get public keys")
+	}
+	if len(keys) == 0 {
+		return errors.New("no keys for user with email: " + u.Email)
+	}
```

Comment:
This should not be an error.

The other package types are signed with a generated per-user key. I see no need to verify an upload. If an user is authenticated on the instance it should be enough.

Reply:
I think there should be a way to upload a package with a script or client that uses GnuPG, since arch packages are already signed with gpg, but i can create separate method to make upload using gitea's existing auth system.

I mean it makes sense to use shell script or cli utility when it comes to package uploads, since package delivery is very common operation, and that would be nice if such script/utility would reuse already existing package signature on server side for validation.

Let me know if i should make additional endpoints with standart gitea authorization methods for incoming package validation.